README.md
# faur

A `faur` instance is a mirror of all package data on the AUR. `faur` is...

- **Simple**
  - There is only a single endpoint (i.e. no difference between `info` and `search`).
  - JSON format is identical to the AUR RPC (but failure is always an empty list).
- **Fast**
  - All data is held in memory with custom indices for near-instant lookups.
  - No rate limits.
- **Featureful**
  - Searching by "provides".
  - Searching by multiple terms at once (has "AND" semantics).
- **Small**
  - ~300 lines of Clojure.
  - No external database or other infrastructure required.

For instance, visit:

- <https://faur.fosskers.ca/packages?names=aura&by=prov>
- <https://faur.fosskers.ca/packages?names=nintendo,switch&by=desc>

### API

There is only a single endpoint: `packages`.

| Endpoint                          | Function                                                                         | Big-O Efficiency |
|-----------------------------------|----------------------------------------------------------------------------------|------------------|
| `packages?names=<TOKENS>`         | Look up `m`-many packages by name                                                | `O(mlogn)`       |
| `packages?names=<TOKEN>&by=prov`  | Find packages that satisfy `TOKEN`                                               | `O(logn)`        |
| `packages?names=<TOKENS>&by=desc` | Find packages that contain all `TOKENS` in their names / descriptions / keywords | `O(mlogn)`       |

Where multiple `TOKENS` are accepted, these are separated by commas, as in:

```
packages?names=spotify,teams,zoom
```

**Caveat:** `by=desc` is term-based, not regex based. This is for performance
reasons. So, `packages?names=aura&by=desc` will match on `aura-bin` but not on
`auralcap`.

### Running a `faur` Instance

Running a personal `faur` instance is simple. First, you'll need package data.
From the top-level of the project repo:

```sh
wget https://aur.archlinux.org/packages-meta-ext-v1.json.gz
gzip -d packages-meta-ext-v1.json.gz
```

Then simply:

```sh
clojure -M -m faur
```

This will run a local `faur` server on http://0.0.0.0:8080 . To run in TLS mode,
pass `--key` and `--cert` as well and HTTPS requests will be accepted on port
443. For example, here is how the official faur instance itself is invoked:

``` sh
clojure -M -m faur --key /etc/letsencrypt/live/faur.fosskers.ca/privkey.pem --cert /etc/letsencrypt/live/faur.fosskers.ca/fullchain.pem 
```

### Live Remote REPL

For live debugging, an [nREPL][0] server is embedded and ran on
`localhost:7888`. If running `faur` on a remote server, you can access this
nREPL remotely by first doing an SSH port-forward on your local machine:

``` sh
ssh -NL 7888:localhost:7888 root@<IP-OF-REMOTE-SERVER> -v
```

and then performing a `cider-connect-clj` (or similar), selecting
`localhost:7888` as the target. Once connected, you're free to inspect the
various Atoms or redefine functions.

[0]: https://nrepl.org/nrepl/index.html


CHANGELOG.md
# faur

### Unreleased

Faur has been rewritten Clojure. A separate `faur-supervisor` is no longer necessary.

### 0.2.2 (2022-11-02)

#### Fixed

- Account for new AUR package info fields.

### 0.2.0 (2022-06-15)

##### Changed

- Read JSON dump pulled directly from the AUR, as opposed to homespun YAML.


src/faur.clj
(ns faur
  (:require
   [cheshire.core :as json]
   [clojure.set :as set]
   [clojure.string :as str]
   [clojure.tools.cli :refer [parse-opts]]
   [data-fetch :as fetch]
   [less.awful.ssl :as ssl]
   [nrepl.server :as nrepl]
   [ring.adapter.jetty :as ring]
   [ring.middleware.params :refer [wrap-params]]
   [taoensso.timbre :refer [info set-min-level!]]))

(def cli-options
  [["-p" "--port PORT" "Port number"
    :default 8080
    :parse-fn #(Integer/parseInt %)]
   [nil "--key PATH"  "Path to a TLS certificate's private key"]
   [nil "--cert PATH" "Path to a TLS certificate"]])

(defonce all-packages (atom []))
(defonce by-names (atom {}))
(defonce by-provides (atom {}))
(defonce by-words (atom {}))
(defonce req-count (atom 0))

(defn bad-route
  "Some unrecognized route was called."
  []
  {:status 404
   :headers {"Content-Type" "text/html"}
   :body "That route does not exist."})

(defn success
  "Yield a good response as JSON."
  [edn]
  (swap! req-count inc)
  {:status 200
   :headers {"Content-Type" "application/json"}
   :body (json/generate-string edn)})

(defn find-by-names
  "Yield the packages the match the given names."
  [all-by-name query-pkgs]
  (->> query-pkgs
       (map #(get all-by-name %))
       (filter identity)))

(defn find-by-provs
  "Yield the packages that match the given provides."
  [all-by-name all-by-provs query-pkgs]
  (->> query-pkgs
       (map #(get all-by-provs %))
       (apply set/union)
       (map #(get all-by-name %))))

(comment
  (->> "aura,git"
       (#(str/split % #","))
       (find-by-provs @by-names @by-provides)
       (#(doto % tap>))))

(defn find-by-words
  "Yield the packages that contain the given words."
  [all-by-name all-by-words query-terms]
  (->> query-terms
       (map #(get all-by-words %))
       (apply set/intersection)
       (take 100)
       (map #(get all-by-name %))))

(comment
  (->> #{"nintendo" "switch"}
       (find-by-words @by-names @by-words)
       (map #(select-keys % [:Name :Description]))
       (sort-by :Name)))

(defn handler [request]
  (let [uri    (:uri request)
        method (:request-method request)
        pkgs   (->> (get (:params request) "names")
                    (#(if (nil? %) [] (str/split % #",")))
                    (into #{}))  ; Prevents repeated query terms.
        by     (get (:params request) "by")]
    (if (and (= :get method) (= "/packages" uri))
      (cond (= by "prov") (->> pkgs (find-by-provs @by-names @by-provides) success)
            (= by "desc") (->> pkgs (find-by-words @by-names @by-words) success)
            (nil? by)     (->> pkgs (find-by-names @by-names) success)
            :else         (bad-route))
      (bad-route))))

(defn server-config
  "Configure the REST server based on the given commandline arguments. Will run in
  HTTS-mode if a TLS cert/key are available."
  [{:keys [key cert port]}]
  (let [basic {:port port :join? false}]
    (if (and key cert)
      (assoc basic
             :ssl-context (ssl/ssl-context key cert)
             :ssl-port 443
             :client-auth :want)
      basic)))

(defonce fut (atom nil))
(defonce server (atom nil))

(defn -main [& args]
  (let [opts (:options (parse-opts args cli-options))]
    (set-min-level! :info)
    (info "Reading initial package data...")
    (fetch/refresh-package-data all-packages by-names by-provides by-words)
    (info "Spawing refresh thread...")
    (swap! fut (constantly (future (fetch/update-data all-packages by-names by-provides by-words))))
    (info "Starting API server.")
    (swap! server (constantly (ring/run-jetty (wrap-params #'handler) (server-config opts))))
    (info "Starting nREPL.")
    (nrepl/start-server :port 7888)))
